### ISSUE TITLE
Problem with implicit higher-kind-yielding typelevel computation

### ISSUE THREAD

**Issue by ctongfei** [scala/bug#11274]
```scala
class FloatTensor[A]

trait DType[R] {
  type Tensor[_]
}

object DType {
  type Aux[R, T[_]] = DType[R] { type Tensor[A] = T[A] }

  implicit val float: DType.Aux[Float, FloatTensor] = new DType[Float] {  // ***
    type Tensor[A] = FloatTensor[A]
  }
}

object Bug extends App {

  def f[T[_], R](implicit dType: DType.Aux[R, T]) = ???

  f(dType = DType.float)   // Failure here even if the implicit val is EXPLICITLY given

}
```

This fails compilation by issuing the following message:
```
Error:(27, 3) inferred kinds of the type arguments (nexus.DType.float.Tensor[A],Float) do not conform to the expected kinds of the type parameters (type T,type R).
nexus.DType.float.Tensor[A]'s type parameters do not match type T's expected parameters:
type Tensor has one type parameter, but type T has one
  f(dType = DType.float)
```

This is clearly weird: what does it mean that "type `Tensor` has **one** type parameter, but type `T` has **one**" ???

However, if the `implicit val` on the line marked with `***` is changed to **`implicit def`**, it miraculously compiles!

This is probably related to #10849 but I'm not sure -- the output message here is clearly nonsense.

**Comment by joroKr21** [comment]
If val vs def matters maybe related to singleton types :?

**Comment by ctongfei** [comment]
@joroKr21 I see that you fixed #10849 -- can you confirm that this examples runs under your fix?

**Comment by joroKr21** [comment]
> @joroKr21 I see that you fixed #10849 -- can you confirm that this examples runs under your fix?

Yeah it's fixed on 2.13.x branch - works both implicitly and explicitly. You can easily check in the REPL by cloning this repo, and run the `scala` command in the `sbt` console.

**Comment by ctongfei** [comment]
OK Thanks! Then this is fixed and should be closed?
